addons/stock_account/models/product.py

```diff
@@ -72,6 +110,100 @@ def write(self, vals):
     # -------------------------------------------------------------------------
     # Misc.
     # -------------------------------------------------------------------------
+    def _enable_lot_valuated(self):
+        self.ensure_one()
+        if not self.product_variant_ids.stock_valuation_layer_ids:
+            return
+        move_vals_list = []
+        lot_by_product = defaultdict(lambda: defaultdict(float))
+        neg_lots = self.env['stock.quant']._read_group([
+            ('product_id', 'in', self.product_variant_ids.ids),
+            ('lot_id', '!=', False),
+            ], ['product_id', 'location_id', 'lot_id'], ['quantity:sum'],
+            having=[('quantity:sum', '<', 0)])
+        lots = self.env['stock.quant']._read_group([
+            ('product_id', 'in', self.product_variant_ids.ids),
+            ('lot_id', '!=', False),
+            ], ['product_id', 'location_id', 'lot_id'], ['quantity:sum'],
+            having=[('quantity:sum', '>', 0)])
+        for group in lots:
+            if group[1]._should_be_valued():
+                lot_by_product[group[0]][group[2]] += group[3]
+        for group in neg_lots:
+            if group[1]._should_be_valued():
+                raise UserError(_("Lot %(lot)s has a negative quantity in stock. Correct this \
+                        quantity before enabling lot valuation", lot=group[2]))
+
+        description = _("Enabling lot valuation for product %s.", self.display_name)
+        out_svl_vals_list, p_orig, products = self.env['product.product']._svl_empty_stock(description, product_template=self)
+        svl_vals_list = out_svl_vals_list
+        for variant in products:
+            rounding = variant.uom_id.rounding
+            if float_is_zero(variant.quantity_svl, precision_rounding=rounding):
+                continue
+            price_unit = variant.value_svl / variant.quantity_svl
+            for lot, qty in lot_by_product[variant].items():
+                qty_to_remove = min(qty, p_orig[variant.id])
+                p_orig[variant.id] -= qty_to_remove
+                svl_vals = variant._prepare_in_svl_vals(qty_to_remove, price_unit, lot_id=lot)
```

Comment:
You should take the value on the lot if possible and fallback on product standard price